pkg/blockbuilder/blockbuilder.go

```diff
@@ -239,3 +237,306 @@ func (b *BlockBuilder) getLagForPartition(ctx context.Context, partition int32)
+	lastBlockEnd := blockEnd
+	if lastBlockEnd.Before(state.LastBlockEnd) {
+		lastBlockEnd = state.LastBlockEnd
+	}
+	commit := kadm.Offset{
+		Topic:       commitRec.Topic,
+		Partition:   commitRec.Partition,
+		At:          commitRec.Offset + 1, // offset+1 means everything up to (including) the offset was processed
+		LeaderEpoch: commitRec.LeaderEpoch,
+		Metadata:    marshallCommitMeta(commitRec.Timestamp.UnixMilli(), lastSeenOffset, lastBlockEnd.UnixMilli()),
+	}
+	newState := partitionState{
+		Lag:                   state.Lag - (commit.At - firstRec.Offset), // the new lag is the distance between fully processed offsets
+		Commit:                commit,
+		CommitRecordTimestamp: commitRec.Timestamp,
+		LastSeenOffset:        lastSeenOffset,
+		LastBlockEnd:          lastBlockEnd,
+	}
+	if err := b.commitState(ctx, logger, b.cfg.ConsumerGroup, newState); err != nil {
+		return state, err
+	}
+
+	return newState, nil
+}
+
+func (b *BlockBuilder) commitState(ctx context.Context, logger log.Logger, group string, state partitionState) error {
+	offsets := make(kadm.Offsets)
+	offsets.Add(state.Commit)
+
+	boff := backoff.New(ctx, backoff.Config{
+		MinBackoff: 100 * time.Millisecond,
+		MaxBackoff: time.Second,
+		MaxRetries: 10,
+	})
+	for boff.Ongoing() {
+		err := kadm.NewClient(b.kafkaClient).CommitAllOffsets(ctx, group, offsets)
+		if err == nil {
+			break
+		}
+		level.Warn(logger).Log("msg", "failed to commit offsets; will retry", "err", err, "offset", state.Commit.At)
+		boff.Wait()
+	}
+	if err := boff.ErrCause(); err != nil {
+		return fmt.Errorf("commit with partition %d, offset %d: %w", state.Commit.Partition, state.Commit.At, err)
+	}
+
+	level.Debug(logger).Log("msg", "successfully committed to Kafka", "offset", state.Commit.At)
+
+	return nil
+}
+
+func (b *BlockBuilder) uploadBlocks(ctx context.Context, tenantID, dbDir string, blockIDs []string) error {
+	buc := bucket.NewUserBucketClient(tenantID, b.bucket, b.limits)
+	for _, bid := range blockIDs {
+		blockDir := path.Join(dbDir, bid)
+		meta, err := block.ReadMetaFromDir(blockDir)
+		if err != nil {
+			return fmt.Errorf("read block meta for block %s (tenant %s): %w", bid, tenantID, err)
+		}
+
+		if meta.Stats.NumSamples == 0 {
+			// No need to upload empty block.
+			level.Info(b.logger).Log("msg", "skip uploading empty block", "tenant", tenantID, "block", bid)
+			return nil
+		}
+
+		meta.Thanos.Source = block.BlockBuilderSource
+		meta.Thanos.SegmentFiles = block.GetSegmentFiles(blockDir)
+
+		if meta.Compaction.FromOutOfOrder() && b.limits.OutOfOrderBlocksExternalLabelEnabled(tenantID) {
+			// At this point the OOO data was already ingested and compacted, so there's no point in checking for the OOO feature flag
+			meta.Thanos.Labels[mimir_tsdb.OutOfOrderExternalLabel] = mimir_tsdb.OutOfOrderExternalLabelValue
+		}
+
+		boff := backoff.New(ctx, backoff.Config{
+			MinBackoff: 100 * time.Millisecond,
+			MaxBackoff: time.Second,
```

Comment:
I would significantly increase the backoff here. If there's a temporary networking issue it's unlikely will be resolved in 10s (10 x 1s max backoff). We could have way longer backoffs instead, explaining the reason with a comment.